File: scripts/balance.py
#!/usr/bin/env python3
"""Simulate strategies to check CryptoWarz is still a game.

A tuning change that makes one simple heuristic win every run turns the game
into a formula - which is exactly what the first price model did. Run this
after touching prices, biases, debt or capacity, and read the table: doing
nothing must lose, acting at random must lose badly, and better judgement must
raise both the median and the ceiling.
"""

from __future__ import annotations

import statistics
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from cryptowarz.coins import COINS          # noqa: E402
from cryptowarz.game import DAYS, Game      # noqa: E402
from cryptowarz.stations import STATIONS    # noqa: E402

RUNS = 200


def _liquidate(game: Game) -> None:
    for symbol, holding in list(game.player.wallet.items()):
        if holding.qty > 0:
            game.sell(symbol, holding.qty)


def _cheapest(game: Game) -> str:
    return min(((c.symbol, (game.market.price(c.symbol) - c.low) / (c.high - c.low))
                for c in COINS if c.symbol != "USDC"), key=lambda x: x[1])[0]


def do_nothing(game: Game) -> None:
    pass


def at_random(game: Game) -> None:
    _liquidate(game)
    coin = game.rng.choice(COINS)
    qty = game.max_buyable(coin.symbol)
    if qty > 0:
        game.buy(coin.symbol, qty * 0.9)


def buy_the_dip(game: Game) -> None:
    _liquidate(game)
    symbol = _cheapest(game)
    qty = game.max_buyable(symbol)
    if qty > 0:
        game.buy(symbol, qty * 0.95)


def dip_and_clear_debt(game: Game) -> None:
    _liquidate(game)
    if game.station.has_shark and 0 < game.player.debt < game.player.cash * 0.55:
        try:
            game.repay(game.player.debt)
        except ValueError:
            pass
    if game.station.has_upgrades and game.player.debt == 0 and game.player.cash > 80_000:
        try:
            game.buy_capacity()
        except ValueError:
            pass
    symbol = _cheapest(game)
    qty = game.max_buyable(symbol)
    if qty > 0:
        game.buy(symbol, qty * 0.95)


STRATEGIES = [
    ("do nothing", do_nothing),
    ("buy at random", at_random),
    ("buy the cheapest", buy_the_dip),
    ("+ clear the debt", dip_and_clear_debt),
]


def play(seed: int, strategy) -> float:
    game = Game(seed=seed)
    for _ in range(DAYS):
        if game.finished:
            break
        try:
            strategy(game)
        except (ValueError, KeyError):
            pass
        options = [s.name for s in STATIONS if s.name != game.station.name]
        try:
            game.travel(game.rng.choice(options))
        except ValueError:
            break
    return game.final_score()


def main() -> int:
    print(f"\n  {RUNS} runs per strategy\n")
    print(f"  {'strategy':<20}{'median':>12}{'p10':>12}{'p90':>12}{'best':>13}{'solvent':>10}")
    print("  " + "-" * 79)
    results = {}
    for name, strategy in STRATEGIES:
        scores = sorted(play(i, strategy) for i in range(RUNS))
        results[name] = scores
        solvent = sum(1 for s in scores if s > 0) / len(scores)
        print(f"  {name:<20}{statistics.median(scores):>12,.0f}"
              f"{scores[len(scores) // 10]:>12,.0f}{scores[-len(scores) // 10]:>12,.0f}"
              f"{scores[-1]:>13,.0f}{solvent:>9.0%}")
    print()

    problems = []
    if statistics.median(results["do nothing"]) > 0:
        problems.append("doing nothing is profitable - the debt is too soft")
    if sum(1 for s in results["buy at random"] if s > 0) / RUNS > 0.35:
        problems.append("random play survives too often - the market is too forgiving")
    solvent = sum(1 for s in results["buy the cheapest"] if s > 0) / RUNS
    if solvent > 0.85:
        problems.append(f"one simple heuristic wins {solvent:.0%} of runs - it is a formula, not a game")
    if solvent < 0.15:
        problems.append(f"a sensible strategy only survives {solvent:.0%} of runs - it is punishing, not hard")

    for problem in problems:
        print(f"  BALANCE: {problem}")
    if not problems:
        print("  BALANCE: healthy - nothing wins for free, nothing is hopeless.")
    print()
    return 1 if problems else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: web/balance.js
/* The balance table from scripts/balance.py, run against the JS port.

   The medians cannot match Python exactly - mulberry32 and the Mersenne
   Twister produce different streams - so what is compared is the shape:
   doing nothing must lose, random play must lose badly, and better judgement
   must raise both the survival rate and the ceiling. */
const { Game, STATIONS, COINS, bestChoice } = require("./game.js");

const RUNS = Number(process.argv[2] || 200);
const median = a => { const s = [...a].sort((x, y) => x - y); return s[Math.floor(s.length / 2)]; };

function liquidate(g) {
  for (const [sym, h] of Object.entries(g.player.wallet)) if (h.qty > 0) g.sell(sym, h.qty);
}
function cheapest(g) {
  let best = null;
  for (const c of COINS) {
    if (c.symbol === "USDC") continue;
    const pos = (g.market.prices[c.symbol] - c.low) / (c.high - c.low);
    if (!best || pos < best[1]) best = [c.symbol, pos];
  }
  return best[0];
}

const STRATEGIES = [
  ["do nothing", () => {}],
  ["buy at random", g => {
    liquidate(g);
    const c = g.rng.choice(COINS);
    const q = g.maxBuyable(c.symbol);
    if (q > 0) g.buy(c.symbol, q * 0.9);
  }],
  ["buy the cheapest", g => {
    liquidate(g);
    const s = cheapest(g);
    const q = g.maxBuyable(s);
    if (q > 0) g.buy(s, q * 0.95);
  }],
  ["+ clear the debt", g => {
    liquidate(g);
    if (g.station.shark && g.player.debt > 0 && g.player.debt < g.player.cash * 0.55) {
      try { g.repay(g.player.debt); } catch (e) { /* not affordable */ }
    }
    if (g.station.shop && g.player.debt === 0 && g.player.cash > 80000) {
      try { g.buyCapacity(); } catch (e) { /* not affordable */ }
    }
    const s = cheapest(g);
    const q = g.maxBuyable(s);
    if (q > 0) g.buy(s, q * 0.95);
  }],
];

function play(seed, strategy) {
  const g = new Game(seed);
  for (let d = 0; d < 30; d++) {
    if (g.finished) break;
    try { strategy(g); } catch (e) { /* an unaffordable move is a no-op */ }
    const options = STATIONS.map((s, i) => i).filter(i => STATIONS[i].name !== g.station.name);
    try { g.travel(g.rng.choice(options)); } catch (e) { break; }
    // A standoff blocks every other action until it is answered. A bot that
    // ignored one would not merely skip an event - it would stop playing, and
    // quietly report the rest of the game as unwinnable.
    if (g.pending) { try { g.resolve(bestChoice(g)); } catch (e) { break; } }
  }
  return g.finalScore();
}

const results = {};
for (const [name, fn] of STRATEGIES) {
  const scores = [];
  for (let i = 0; i < RUNS; i++) scores.push(play(i, fn));
  scores.sort((a, b) => a - b);
  results[name] = {
    median: median(scores),
    p90: scores[Math.floor(scores.length * 0.9)],
    best: scores[scores.length - 1],
    solvent: scores.filter(s => s > 0).length / scores.length,
  };
}

if (process.argv.includes("--json")) {
  process.stdout.write(JSON.stringify(results));
} else {
  console.log(`\n  JS port, ${RUNS} runs per strategy\n`);
  console.log("  " + "strategy".padEnd(20) + "median".padStart(12) + "p90".padStart(12)
              + "best".padStart(13) + "solvent".padStart(10));
  console.log("  " + "-".repeat(67));
  for (const [name, r] of Object.entries(results)) {
    console.log("  " + name.padEnd(20)
      + Math.round(r.median).toLocaleString().padStart(12)
      + Math.round(r.p90).toLocaleString().padStart(12)
      + Math.round(r.best).toLocaleString().padStart(13)
      + (Math.round(r.solvent * 100) + "%").padStart(10));
  }
  console.log();
}
